README.md
#COIAS_programの使い方

##下準備

0.0. (Readme_COIAS2.pdfも参照してください) 3系のpythonのインストール。anaconda(pythonの統合開発環境)をインストールした方が良い。その上で以下のパッケージなどをインストールする: numpy, scipy, matplotlib, astropy, ephem, cython, pandas, pillow, photutils, SExtractor, astroquery, julian

0.1. より簡単な方法: 3系のpythonがインストールされたanacondaをインストールしたのち, $bash env_setting.sh でほぼ環境構築が終わる. ただし環境によってはうまくインストールできない or 使えないパッケージがあったりするので, それらは手でインストールする. coiasという仮想環境上に環境を構築するので, シェルを開き直した後は $source activate coias で環境を切り替えれば使用可能. => この方法でインストールした人は下準備5.までの全ての準備をスキップして実行方法から開始できる.

1. githubからこのプログラム一式をダウンロードして、任意のディレクトリに置く。本Readmeではこのディレクトリを /COIAS_program_path と呼ぶ。
2. /COIAS_program_path以下の全てのpythonスクリプトとシェルスクリプトにchmodで実行許可を与えておく。
3. シェルの環境変数PATHに、このディレクトリへのパス (/COIAS_program_path) とfindOrbへのパス (/COIAS_program_path/findOrb) を追加する。使用するシェルはbashが前提のようなので、bashでパスを通す。(以前COIASのプログラム群を使用しており昔のパスが残っている場合は、使用したいプログラム群へのパスのみが環境変数PATHに登録されるように注意する)
4. cythonのビルド。/COIAS_program_path にターミナルで移動して、 python setup12.py build_ext --inplace と入力する。
5. findOrbのコンパイル。/COIAS_program_path/findOrb にターミナルで移動して、以下のコマンドを打つことでコンパイルを実行する。デフォルトのコンパイラは g++ なので、無い場合はインストールするか、持っているc++用のコンパイラを linlunar.make, linmake ファイル中の CC=コンパイラ名 に指定する必要がある。
   1. もし一度コンパイルしたことがあったら、rm dos_find, rm lunar.a, rm *.o ですでにある実行ファイル達を削除する
   2. make -f linlunar.mak
   3. make -f linmake

##実行方法
1. 任意の好きなディレクトリに5枚のwarp画像 (warp-*.fits) を用意して、ターミナルでこのディレクトリに移動しておく。ほぼ全ての中間ファイルやpngファイルはこのディレクトリ (カレントディレクトリ) に展開される。
2. AstsearchR と打ち込んで同スクリプトを使用し、binning、マスク画像引き、光源検出、視野周辺の既知小惑星取得、移動天体検出、測光、pngファイル生成、などなどを実行する。
3. COIAS.py と打ち込んで、GUIで移動天体を目視で確認する。操作の詳細は省略するが、png画像を選び、blinkさせて移動天体だと思うものの四角を左クリックして選択し、選択が終わったらoutputを押してmemo.txtを出力させる。
4. prempedit と打ち込んで、MPCフォーマットに再整形。
5. prempedit3.py [新天体の通し番号] と打ち込んで、名前の付け替え。第二引数に今まで自分が見つけた新天体の番号のうち一番大きいもの+1を指定する。
6. redisp と打ち込んでCOIASを再表示する準備。
7. ReCOIAS.py を実行するが、プログラムを準備していないので省略。ReCOIAS.pyはCOIAS.pyで読み込むdisp.txtをredisp.txtに変えただけである。つまり、1度目のCOIAS.pyで選んだ天体のみに四角をつけて確認するだけ。
8. mpc4.txtを複製してsend_mpc.txtに名前を書き換える。手作業。何回か測定した場合は、複数回の測定でできた複数のmpc4.txtの内容をまとめてsend_mpc.txtに記入しても良い。
9. AstsearchR_afterReCOIAS と打ち込んで同スクリプトを使用し、重複行の削除、findOrbを用いた軌道測定、誤差が大きいデータの削除、新天体に米印をつける、などを実行する。
10. 作成されたsend_mpc.txtが完成形で、MPCに送信する報告メールのデータ部分になる。

##たまにするべきこと
1. (たまに実行して最新のMPCのデータベースを取り込んでおく。 初回はスクリプト中で自動で実行されるので、しなくて良い) ターミナルで getMPCORB_and_mpc2edb と打ち込んで同スクリプトを実行することで、 MPCからMPCORB.DATを~/.coiasにダウンロードし、さらに解析してedb形式に書き換える。~/.coiasは初回の実行時に自動で作られる隠しディレクトリ。


CHANGES.txt
2021-07-15 Natsuho Maeda 
* Making a file (CHANGES.txt) to record change logs.

2021-07-16 Keisuke Sugiura
* Commit COIAS2.py that includes manual photometry as COIAS.py

2021-07-20 Keisuke Sugiura
* Change name of binned files warp-*_bin.fits -> warpbin-*.fits

2021-07-26 Keisuke Sugiura
* Modify COIAS.py so that we can cancel the selection of checked object by clicking again

2021-08-10 Natsuho Maeda
* Modify header copy part in binning2R.py. Modifyed correction factor of 2 to nbin for keywords "CRPIX1", "CRPIX2", "CD1_1", "CD1_2", "CD2_1", "CD2_2".
* Modify name of images which are displayed on COIAS. Before: warp[1-5]_bin.png  After: [1-5]_disp-coias.png. Modyfied scripts: subm2.py, COIAS.py (## Only for .png name. Not changed for .fits and .dat file name.)
* Modify the number shift between asteroid coordinates and image number at COIAS.py

2021-8-22 Seitaro Urakawa
*Modify mktraclet.pyx

2021-8-28 Seitaro Urakawa
*Fix bugs of mpc format. Modify mpc1b.py, mpc2b.py, mpc1c.py, mpc2c.py, astsearch_new,and astsearch_manual.

2021-10-19 Seitaro Urakawa
*Commit and push COAS4.py(+manual measurements version)

2021-10-20 Seitaro Urakawa
*cp COIAS4.py COIAS_m.py

2021-10-24 Seitaro Urakawa
mv COIAS.py COIAS_211024.py
mv COIAS_m.py COIAS.py
*Commit and push COIAS_211024.py (previous without manual mode version) and COIAS.py (+ manual mode. latest version)

2021-10-28 Seitaro Urakawa
Fix bug of manual mode. Revised version COIAS_211026.py
cp COIAS_211026.py COIAS.py
Revised ReCOIAS.py
*Commit and push COIAS_211026.py (backup of COIAS.py), COIAS.py (+latest version), and ReCOIAS.py

Revised binnig2R.py and binning3R.py (Correspondence both fits header keyword 'TIME-MID' and 'DATE-AVG')
Add the execution permission (mpc1c.py, mpc2c.py, match2E.py, astsearch1Q5.py, astsearch_manual, binning3R.py, and ReCOIAS.py)
*Commit and push mpc1c.py, mpc2c.py, match2E.py, astsearch1Q5.py, astsearch_manual, binning3R.py, and ReCOIAS.py

2021-10-30 Seitaro Urakawa
Modify astserch_manual for display of  manual detection object, and astsearch1Q5.py
Modify ReCOIAS.py(window of x,y pixel is widen.)
Push "How_to_use_manual_mode.pdf"

2021-11-9 Seitaro Urakaw
Fix bug of match2E.py (add correct time information).
Modify Astsearch_afterReCOIAS and komejirushi.py. Change output file name from mpcOeg.txt to send_mpc.txt

2021-12-05 Natsuho Maeda
Fix bug of getinfo_numbered2D.py (bug that occurs when the asteroid name contains "'".)

2021-12-21 Natsuho Maeda
Modify prempsearchC (Change the criteria of division into numbered or karifugo.)

2021-12-24 Natsuho Maeda
Combine binning2R.py and binng3R.py to binning.py. Add a function to check existence of FLUXMAG0 in header.

2022-1-14 Natsuho Maeda
Modify getinfo_numbered2D.py (change data type of tmp7 into numpy.ndarray when declaring)

2022-4-6 Keisuke Sugiura
Add env_setting.sh that was produced by Maeda-san and modify README.md

2022-4-7 Keisuke Sugiura
change place of all files in .coias/ to .coias/param/

2022-4-14 Keisuke Sugiura
Modify some scripts so that we store all precise orbits of known bodies and reuse them

2022-4-20 Keisuke Sugiura
Modify some scripts regarding changing data to MPC format

2022-4-20 Keisuke Sugiura
Modify some scripts so that we include bright known asteroids to the list of known asteroids

COIAS.py
#!/usr/bin/env python3
# -*- coding: UTF-8 -*
#Asthunter ver 0.2 /2020/1/27"
#Asthunter => COIAS ver 0.0 /2020/6/10"

import tkinter as tk
import tkinter.filedialog as fd
import sys, os.path
import os
from PIL import Image
from tkinter import ttk
from time import sleep
import numpy as np
import re
import tkinter.scrolledtext as st
#SU. revised
from tkinter import messagebox
import tkinter.simpledialog as simpledialog
#input ast data

ast_xy = np.loadtxt("disp.txt",dtype = 'str')
#img = Image.open('warp1_bin.png')
img = Image.open('1_disp-coias.png') # NM modified 2021-08-10
xpix = img.size[0]
ypix = img.size[1]
path_name = os.getcwd()



#class
class Asthunter(tk.Frame):
    def __init__(self, master=None):        
        tk.Frame.__init__(self,master)
        self.grid()
        self.create_widgets()
#first window
    def create_widgets(self):
        self.button1 = tk.Button(root,text='Load img',font=("",14),command = self.load_file)
        self.button1.grid(row=0,column =0,sticky = 'NW')
        self.button4 = tk.Button(root,text='Quit',font=("",14),command = root.quit)
        self.button4.grid(row=0,column=2,sticky = 'NE')
        self.button5 = tk.Button(root,text='OutPut',font=("",14),command = self.output)
        self.button5.grid(row=0,column=1,sticky = 'NE')
#moving object H number 
        self.t0 = st.ScrolledText(font=("Helvetica", 14),width = 10,height = 20)
        self.t0.grid(row = 1,column =0,columnspan = 1000,sticky = tk.W+tk.E)
        
#second window
    def sub_window(self):
        global image_data
        print("filenums",self.filenums)
#        global inputnumber
#        self.inputnumber = []
        self.file_num = 0
        self.SqOnOffFlag = 1
        self.sub_win = tk.Toplevel(root)
        self.sub_win.title('COIAS ver 0')
        self.sub_win.geometry('1440x900')
        self.canvas = tk.Canvas(self.sub_win,width=1350,height = 800)
        self.canvas.grid(row=1,column=0,columnspan = 70, sticky=tk.W+tk.E+tk.N+tk.S)
#set scroll
        self.xscroll = tk.Scrollbar(self.sub_win,orient = tk.HORIZONTAL,command = self.canvas.xview)
        self.xscroll.grid(row=2,column=0,columnspan = 70,sticky = tk.E+tk.W)
        self.yscroll = tk.Scrollbar(self.sub_win,orient = tk.VERTICAL,command = self.canvas.yview)
        self.yscroll.grid(row=1,column=70,sticky = tk.N+tk.S)
        self.canvas.config(yscrollcommand = self.yscroll.set)
        self.canvas.config(xscrollcommand = self.xscroll.set)

        #   scrollregion (west,north,east,south)
        self.canvas.config(scrollregion = (0,0,xpix,ypix))
        
        # set first image (center x,y coord of image)/self.image_on_canvas is tag
        self.image_on_canvas = self.canvas.create_image(xpix/2,ypix/2,image = self.image_data[self.image_number])
        
        # set ast coord of the first image/self.coord_on_canvas is tag
        self.astdata = [ast_xy[i] for i in np.where(ast_xy[:,1] == str(self.filenums[self.image_number] - 1))] # NM 2021.08.10
        
        for i in range(len(self.astdata[0])):
            xpos = int(float(self.astdata[0][i][2]))
            ypos = int(float(self.astdata[0][i][3]))
            #set moving object personal name
            name = str(self.astdata[0][i][0])
            
            self.coord_on_canvas = self.canvas.create_rectangle(xpos-20,ypix-ypos+20,xpos+20,ypix-ypos-20,outline='#000000',width=5)
            self.canvas.create_text(xpos-25,ypix-ypos-30,text = name,fill='#000000',font=("Purisa",16))

#set condition
        self.do_blink = False
#set button
        self.button2 = tk.Button(self.sub_win,text='Blink start/stop',font=("",14),command = self.start_stop_blinking)
        self.button2.grid(row=0,column =0,sticky = tk.W)
        self.button3 = tk.Button(self.sub_win,text='close window',font=("",14),command=self.sub_win.destroy)
        self.button3.grid(row=0,column =69,sticky = tk.E)
        self.button_back = tk.Button(self.sub_win, text="Back",command = self.onBackButton,font=("",14))
        self.button_back.grid(row=0,column=1,sticky = tk.W)
        self.button_next = tk.Button(self.sub_win, text="Next",command = self.onNextButton,font=("",14))
        self.button_next.grid(row=0,column=2,sticky = tk.W)
        self.button_SqOnOff = tk.Button(self.sub_win, text="Sq. On/Off",command = self.sq_on_off,font=("",14))
        self.button_SqOnOff.grid(row=0,column=3,sticky = tk.W)
#set entry of image name
        self.message_num = tk.Entry(self.sub_win,font=("",14),width=10)
        self.message_num.insert(tk.END,(("Image # ") + str(self.image_number +1)))
        self.message_num.grid(row=0, column=4,sticky = tk.W)
#print coord & set entry
        self.canvas.bind('<Motion>', self.mousecoord)
        self.xycoord = tk.Entry(self.sub_win,font=("",14),width=25)
        self.xycoord.insert(tk.END,("X pix, Y pix:"))
        self.xycoord.grid(row=0, column=5, columnspan=20,sticky = tk.W)
#set mouse coordinate at click. If the coordinate is included in the range of new objects, print those numbers to the ScrolledText
        self.canvas.bind('<ButtonPress-1>', self.mouseClickAndOutputObjectsNumber)
#set scrolledText box for message
        self.messageBox = tk.Entry(self.sub_win,font=("", 14), width=50)
        self.messageBox.insert(tk.END,"message:")
        self.messageBox.grid(row=0, column=26, columnspan=42,sticky = tk.W)
#SU added 2021/7/12
#set mouse coorinate by rightclick.=> by spacebar (S.U 2021/9/9 )
        self.sub_win.bind('<KeyPress-space>',self.spacebar)


# load_file        
    def load_file(self):
        self.image_data =[]
        self.filenums = [] # NM added 2021-08-10
#multi select
        self.filename = fd.askopenfilenames(filetypes = [('Image Files', ('.gif', '.png', '.ppm')),
                                               ('GIF Files', '.gif'),
                                               ('PNG Files', '.png'),
                                               ('PPM Files', '.ppm')],
                                  initialdir = path_name)
        n = len(self.filename)
        for self.i in range(0,n):
            self.image_data.append(tk.PhotoImage(file = self.filename[self.i]))
            self.filenums.append( int(self.filename[self.i].split('/')[-1][0]) )  # NM added 2021-08-10
        self.image_number = 0
#make sub_window
        self.sub_window()

#------------------------------------

    def onBackButton(self):
        # 最後の画像に戻る
        if self.image_number == 0:
            self.image_number = len(self.image_data) - 1
        else:
            # 一つ戻る
            self.image_number -= 1

        # 表示画像を更新
        self.draw()

        # Entryの中身を更新
        self.message_num.delete(0, tk.END)
        self.message_num.insert(tk.END, ("Image # " + str(self.image_number+1)))

    def onNextButton(self):
        # 一つ進む
        self.image_number += 1

        # 最初の画像に戻る
        if self.image_number == len(self.image_data):
            self.image_number = 0
        
        # 表示画像を更新
        self.draw()
            
        # Entryの中身を更新
        self.message_num.delete(0, tk.END)
        self.message_num.insert(
            tk.END, ("Image # " + str(self.image_number+1)))

    def start_stop_blinking(self):
        if self.do_blink:
            self.do_blink = False
        else:
            self.do_blink = True
            self.blink()

    def blink(self):
        if self.do_blink:
            self.image_number +=1
         # 最初の画像に戻る
            if self.image_number == len(self.image_data):
                self.image_number = 0

        # 表示画像を更新
            self.draw()
                    
            self.after(100,self.blink)
        # Entryの中身を更新
        self.message_num.delete(0, tk.END)
        self.message_num.insert(
            tk.END, ("Image # " + str(self.image_number+1)))

    def stop(self):
        print('stop')
#--------------------------------------------------------------------
    def mousecoord(self,event):
       # get the top left coord in image
        xp = self.xscroll.get()
        yp = self.yscroll.get()
#top left coord in image
        xp1 = int(xp[0]*xpix)
        yp1 = int(yp[0]*ypix)
#add the mouse position
        self.xp2 = event.x + xp1
        self.yp2 = event.y + yp1
        self.xycoord.delete(0,tk.END)
        self.xycoord.insert(tk.END,("X pix "+str(self.xp2)+", Y pix " +str(self.yp2)))

#SU added 2021/07/12  get coordinate by right click => by spacebar(2021/9/9)
#In the case of spacebar, yp2 needs offset of 30pix. SU don't know the reason(2021/10/28) -------------------------------------------
    def spacebar(self,event):
#        print('right click')
        res = messagebox.askquestion("New number","Same object with a previous image?")
#        print(res)
        if res == "no":
#get the top left coord in image
            xp = self.xscroll.get()
            yp = self.yscroll.get()
#top left coord in gui image
            xp1 = int(xp[0]*xpix)
            yp1 = int(yp[0]*ypix)
            print(xp[self.image_number],yp[self.image_number],xpix,ypix,xp1,yp1,event.x,event.y)
#add the mouse position
            self.xp2 = event.x + xp1
#            self.yp2 = event.y + yp1 
            self.yp2 = event.y + yp1 -30
            self.coord = self.xp2,self.yp2
            print(self.coord)
#            comment = str("If you measure (X, Y) = (") + str(self.xp2) + " pix, " + str(self.yp2) + str(" pix), please input temporay number")
            comment = "Input temporary number"
            self.inputnumber = simpledialog.askstring("Temporary number",comment)

#get filename(full path)
            tmp = str(self.filename[self.image_number])
#only filename
#       tmp[-13:]
#change to fits file name
            if ('nonmask' in tmp):
                tmp2 = 'warp'+tmp[-24]+'_bin.fits'
#                print(tmp2)
            else:
                tmp2 = 'warp'+tmp[-16]+'_bin.fits'
#                print(tmp2)
#out put file on the current directry
#coord + filename
            self.coord = str(self.xp2),str(self.yp2),str(tmp2),str(self.inputnumber)
#        self.coord2 = list(self.coord)
#        self.f2=open(tmp4,'a')
            self.f2 = open('memo2.txt','a')
            self.f2.writelines(str(self.coord)+'\n')
            self.f2.close()
        else:
#get the top left coord in image
            xp = self.xscroll.get()
            yp = self.yscroll.get()
#top left coord in image
            xp1 = int(xp[0]*xpix)
            yp1 = int(yp[0]*ypix)
#add the mouse position
            self.xp2 = event.x + xp1
            self.yp2 = event.y + yp1 - 30
            self.coord = self.xp2,self.yp2
#            print(self.coord)

#get filename(full path)
            tmp = str(self.filename[self.image_number])

#change to fits file name
            if ('nonmask' in tmp):
#                tmp2 = tmp[-24:].replace("png","fits")
                tmp2 = 'warp'+tmp[-24]+'_bin.fits'
            else:
                tmp2 = 'warp'+tmp[-16]+'_bin.fits'
#out put file on the current directry
#coord + filename
            self.coord = str(self.xp2),str(self.yp2),str(tmp2),str(self.inputnumber)
#        self.coord2 = list(self.coord)
#        self.f2=open(tmp4,'a')
            self.f2 = open('memo2.txt','a')
            self.f2.writelines(str(self.coord)+'\n')
            self.f2.close()
           
#------------------------------------------------------------

    def output(self):
        self.f=open('memo.txt','w')
        self.f.write(self.t0.get('1.0','end -1c'))
        #self.f.write(self.t0.get('1.0',tk.END))
        self.f.close()

#KS added 2020/12/24----------------------------------------------------------------
    def mouseClickAndOutputObjectsNumber(self, event):
#get numbers already input in the ScrolledText
        self.inputText = self.t0.get('1.0','end -1c')
        if len(self.inputText)<2:
            self.writtenNumbersText=[]
        else:
            self.writtenNumbersText = self.inputText.split()
        self.writtenNumbers = [int(text) for text in self.writtenNumbersText if text.isdigit()]
        
#get the top left coord in image
        xp = self.xscroll.get()
        yp = self.yscroll.get()
#top left coord in image
        xp1 = int(xp[0]*xpix)
        yp1 = int(yp[0]*ypix)
#add the mouse position
        self.xp2 = event.x + xp1
        self.yp2 = event.y + yp1
        
#get asteroids coordinates stored in disp.txt
        self.astdata = [ast_xy[i] for i in np.where(ast_xy[:,1] == str(self.filenums[self.image_number] - 1))] # NM 2021.08.10
#compare asteroids coordinates and clicked coordinates
        self.matchAsteroidNamesStr = []
        for i in range(len(self.astdata[0])):
            xpos = int(float(self.astdata[0][i][2]))
            ypos = int(float(self.astdata[0][i][3]))
            if xpos-20<self.xp2 and xpos+20>self.xp2 and ypix-(ypos+20)<self.yp2 and ypix-(ypos-20)>self.yp2:
                asteroidNameStr = self.astdata[0][i][0]
                if len(asteroidNameStr)!=7 or asteroidNameStr[0]!='H':
                    #print("This is not a new object! name=",asteroidNameStr)
                    self.messageBox.delete(0, tk.END)
                    self.messageBox.insert(tk.END,"message: not a new object! name="+asteroidNameStr)
                else:
                    self.matchAsteroidNamesStr.append(asteroidNameStr.lstrip("H"))

#put clicked asteroid number that are not yet written on the ScrolledText
        for asteroidNumberStr in self.matchAsteroidNamesStr:
            matchFlag=0
            rowWrittenNumbers=0
            for num in self.writtenNumbers:
                rowWrittenNumbers += 1
                if int(asteroidNumberStr)==num:
                   self.t0.delete("{}.0".format(rowWrittenNumbers),"{}.end +1c".format(rowWrittenNumbers)) #KS added 2021/7/25
                   self.messageBox.delete(0, tk.END)
                   self.messageBox.insert(tk.END,"message:")
                   matchFlag=1
            if matchFlag==0:
                   self.t0.insert(tk.END,asteroidNumberStr+"\n")
                   self.messageBox.delete(0, tk.END)
                   self.messageBox.insert(tk.END,"message:")
                   
        #再描画
        self.draw()


    def sq_on_off(self):
        if self.SqOnOffFlag == 1:
            self.SqOnOffFlag = 0
        elif self.SqOnOffFlag == 0:
            self.SqOnOffFlag = 1

        #再描画
        self.draw()

    def draw(self):
        #get numbers already input in the ScrolledText
        self.inputText = self.t0.get('1.0','end -1c')
        if len(self.inputText)<2:
            self.writtenNumbersText=[]
        else:
            self.writtenNumbersText = self.inputText.split()
        self.writtenNumbers = [int(text) for text in self.writtenNumbersText if text.isdigit()]

        # 表示画像を更新
        self.canvas.delete("all")
        self.image_on_canvas = self.canvas.create_image(xpix/2,ypix/2,image = self.image_data[self.image_number])
        # refresh new ast coord
        self.astdata = [ast_xy[i] for i in np.where(ast_xy[:,1] == str(self.filenums[self.image_number] - 1))] # NM 2021.08.10
        
        for i in range(len(self.astdata[0])):
            xpos = int(float(self.astdata[0][i][2]))
            ypos = int(float(self.astdata[0][i][3]))

            if self.SqOnOffFlag==1:
                name = str(self.astdata[0][i][0])
                matchFlag = 0
                if name.lstrip("H").isdigit():
                    nameInt = int(name.lstrip("H"))
                    for n in self.writtenNumbers:
                        if n == nameInt:
                            matchFlag = 1

                if matchFlag == 0:
                    self.coord_on_canvas = self.canvas.create_rectangle(xpos-20,ypix-ypos+20,xpos+20,ypix-ypos-20,outline='#000000',width=5)   
                    self.canvas.create_text(xpos-25,ypix-ypos-30,text = name,fill='#000000',font=("Purisa",16))
                elif matchFlag == 1:
                    self.coord_on_canvas = self.canvas.create_rectangle(xpos-20,ypix-ypos+20,xpos+20,ypix-ypos-20,outline='#FF0000',width=5)   
                    self.canvas.create_text(xpos-25,ypix-ypos-30,text = name,fill='#FF0000',font=("Purisa",16))
#-----------------------------------------------------------------------------------

root = tk.Tk()
#instance
app = Asthunter(master=root)
app.mainloop()
